src/x/debug/triggering_profile.go

```diff
@@ -0,0 +1,212 @@
+type ContinuousFileProfileOptions struct {
+	FilePathTemplate  string
+	ProfileName       string
+	ProfileDuration   time.Duration
+	ProfileDebug      int
+	Conditional       func() bool
+	Interval          time.Duration
+	InstrumentOptions instrument.Options
+}
+
+// ContinuousFileProfilePathParams is the params used to construct
+// a file path.
+type ContinuousFileProfilePathParams struct {
+	ProfileName string
+	UnixTime    int64
+}
+
+// NewContinuousFileProfile returns a new continuous file profile.
+func NewContinuousFileProfile(
+	opts ContinuousFileProfileOptions,
+) (*ContinuousFileProfile, error) {
+	if opts.FilePathTemplate == "" {
+		return nil, errNoFilePathTemplate
+	}
+	if opts.ProfileName == "" {
+		return nil, errProfileName
+	}
+	if opts.Conditional == nil {
+		opts.Conditional = defaultConditional
+	}
+	if opts.Interval == 0 {
+		opts.Interval = defaultInterval
+	}
+	if opts.InstrumentOptions == nil {
+		return nil, errNoInstrumentOptions
+	}
+
+	tmpl, err := template.New("fileName").Parse(opts.FilePathTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ContinuousFileProfile{
+		filePathTemplate: tmpl,
+		profileName:      opts.ProfileName,
+		profileDuration:  opts.ProfileDuration,
+		profileDebug:     opts.ProfileDebug,
+		conditional:      opts.Conditional,
+		interval:         opts.Interval,
+		logger:           opts.InstrumentOptions.Logger(),
+	}, nil
+}
+
+// Start will start the continuous file profile.
+func (c *ContinuousFileProfile) Start() error {
+	if c.closeCh != nil {
+		return errAlreadyOpen
+	}
+
+	c.closeCh = make(chan struct{})
+	go c.run()
+	return nil
+}
+
+// Stop will stop the continuous file profile.
+func (c *ContinuousFileProfile) Stop() error {
+	if c.closeCh == nil {
+		return errNotOpen
+	}
+
+	close(c.closeCh)
+	c.closeCh = nil
+
+	return nil
+}
+
+func (c *ContinuousFileProfile) run() {
+	closeCh := c.closeCh
+	ticker := time.NewTicker(c.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-closeCh:
+			return
+		case <-ticker.C:
+			if !c.conditional() {
+				continue
+			}
+
+			err := c.profile()
+			if err != nil {
+				c.logger.Error("continuous profile error",
+					zap.String("name", c.profileName),
+					zap.Int("debug", c.profileDebug),
+					zap.Duration("interval", c.interval),
+					zap.Error(err))
+			}
+		}
+	}
+}
+
+func (c *ContinuousFileProfile) profile() error {
+	filePathBuffer := bytes.NewBuffer(nil)
+	filePathParams := ContinuousFileProfilePathParams{
+		ProfileName: c.profileName,
+		UnixTime:    time.Now().Unix(),
+	}
+	err := c.filePathTemplate.Execute(filePathBuffer, filePathParams)
+	if err != nil {
+		return err
+	}
+
+	w, err := os.Create(filePathBuffer.String())
+	if err != nil {
+		return err
+	}
+
+	success := false
+	defer func() {
+		if !success {
+			_ = w.Close()
+		}
+	}()
+
+	switch c.profileName {
+	case ContinuousCPUProfileName:
+		if err := pprof.StartCPUProfile(w); err != nil {
+			return err
+		}
+		time.Sleep(c.profileDuration)
```

Comment:
Is there an issue where the profileDuration is longer than the tick interval and we kick off more than one profile at a time?

Reply:
Thankfully not, higher level loop waits for this method to finish 👍